File: 144.binary_tree_preorder_traversal.py
"""
144. Binary Tree Preorder Traversal
Easy
7.7K
196
Companies
Given the root of a binary tree, return the preorder traversal of its nodes' values.

 

Example 1:


Input: root = [1,null,2,3]
Output: [1,2,3]
Example 2:

Input: root = []
Output: []
Example 3:

Input: root = [1]
Output: [1]
 

Constraints:

The number of nodes in the tree is in the range [0, 100].
-100 <= Node.val <= 100
 
 """
 # Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]: 
        ans = []
        self.r_f(root, ans)
        return ans
    def r_f(self, root, ans):  
        if root:
            ans.append(root.val)
            self.r_f(root.left, ans)
            self.r_f(root.right, ans)
        return 

File: C_programming/144.binary_tree_preorder_traversal.c
/*
144. Binary Tree Preorder Traversal
Solved
Easy
Topics
premium lock iconCompanies

Given the root of a binary tree, return the preorder traversal of its nodes' values.



Example 1:

Input: root = [1,null,2,3]

Output: [1,2,3]

Explanation:

Example 2:

Input: root = [1,2,3,4,5,null,8,null,null,6,7,9]

Output: [1,2,4,5,6,7,3,8,9]

Explanation:

Example 3:

Input: root = []

Output: []

Example 4:

Input: root = [1]

Output: [1]



Constraints:

    The number of nodes in the tree is in the range [0, 100].
    -100 <= Node.val <= 100



Follow up: Recursive solution is trivial, could you do it iteratively?
*/
#include <stdlib.h>
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void tree_size(struct TreeNode *node, int *size){
    if (node == NULL) return;
    (*size)++;
    tree_size(node->left, size);
    tree_size(node->right, size);
}
void traverse(struct TreeNode *node, int *arr, int *index) {
    if (node == NULL) return;
    arr[(*index)++] = node->val;
    traverse(node->left, arr, index);
    traverse(node->right, arr, index);
}
int* preorderTraversal(struct TreeNode* root, int* returnSize) {
    int size = 0;
    tree_size(root, &size);
    *returnSize = size;
    //printf("tree size %d\n", size);
    int *arr = malloc(size * sizeof(int));
    int index = 0;
    traverse(root, arr, &index);
    //for (int i =0; i<size; i++) {
    //    printf("%d:[%d]  ", i, arr[i]);
    //}
    return arr;
}
